nixos/modules/services/web-apps/photoprism.nix

```diff
@@ -0,0 +1,291 @@
+      type = types.bool;
+      default = false;
+      description = ''
+        Flag photos as private that MAY be offensive.
+      '';
+    };
+
+    allowNsfw = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        Allow uploads that MAY be offensive.
+      '';
+    };
+
+    experimental = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Enable experimental features.
+      '';
+    };
+
+    originalsLimit = mkOption {
+      type = types.int;
+      default = 5000;
+      description = ''
+        File size limit for originals in MB (increase for high-res video).
+      '';
+    };
+
+    public = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        No authentication required (disables password protection).
+      '';
+    };
+
+    readonly = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Do not modify originals directory (reduced functionality).
+      '';
+    };
+
+    originalsDir = mkOption {
+      type = types.str;
+      example = "/mnt/nas/photos";
+      description = "Path to your photos.";
+    };
+
+    dataDir = mkOption {
+      type = types.path;
+      default = "/var/lib/photoprism";
+      description = "Data directory for photoprism.";
+    };
+
+    siteTitle = mkOption {
+      type = types.str;
+      default = "PhotoPrism";
+      description = "Web UI title.";
+    };
+
+    siteCaption = mkOption {
+      type = types.str;
+      default = "Browse Your Life";
+      description = "Web UI caption.";
+    };
+
+    extraEnv = mkOption {
+      type = types.attrsOf types.str;
+      default = { };
+      example = { PHOTOPRISM_DEBUG = "true"; };
+      description = ''
+        Additional environment variables for the photoprism service.
+
+        See <link xlink:href="https://dl.photoprism.app/docker/docker-compose.yml"/>
+        and <link xlink:href="https://github.com/photoprism/photoprism/blob/develop/internal/config/flags.go"/>
+        for the supported environment variables.
+      '';
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.photoprism;
+      defaultText = "pkgs.photoprism";
+      description = "Photoprism package to use.";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.databaseDriver == "sqlite" -> (cfg.databaseDsn == null);
+        message = "SQLite is an internal database, databaseDsn must be null";
+      }
+      {
+        assertion = cfg.databaseDriver == "mysql" -> (cfg.databaseDsn != null);
+        message = "databaseDsn must be provided when using mysql databases";
+      }
+    ];
+
+    networking.firewall = lib.mkIf cfg.openFirewall {
+      allowedTCPPorts = [ cfg.port ];
+    };
+
+    users = {
+      users.${cfg.user} = {
+        inherit (cfg) group;
+        home = cfg.dataDir;
+        createHome = true;
```

Comment:
Why do this user needs a home 🤔 ?

Reply:
Photoprism stores data (including a sqlite database) in this directory

Reply:
![image](https://user-images.githubusercontent.com/5727578/172423126-5fad4a6d-3108-480c-9eb9-fa54516ac6c8.png)

Reply:
Yup. `dataDir`'s default value is `/var/lib/photoprism`. I think quite a few modules use this pattern (making the chosen data directory be the service users' home), is it a bad practice?
![image](https://user-images.githubusercontent.com/5727578/172486024-cf8c4cb8-557d-4af0-b863-66e7607eb6a6.png)

Reply:
We should use
```
DynamicUser = true;
User = "photoprism";
StateDirectory = "photoprism";
```

Reply:
Actually that might be problematic with `originalsDir` but that could be cirumvented with `ReadOnlyPaths`.